From 1271c4b3437a8bcf9b35ae8ed03be3dc758fe756 Mon Sep 17 00:00:00 2001
From: Hans de Goede <hdegoede@redhat.com>
Date: Fri, 22 Jun 2012 09:43:08 +0200
Subject: [PATCH] usb-redir: Correctly handle the usb_redir_babble usbredir
 status

Signed-off-by: Hans de Goede <hdegoede@redhat.com>
---
 hw/usb/redirect.c |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/hw/usb/redirect.c b/hw/usb/redirect.c
index 5f55d78..c6358c0 100644
--- a/hw/usb/redirect.c
+++ b/hw/usb/redirect.c
@@ -1058,6 +1058,8 @@ static int usbredir_handle_status(USBRedirDevice *dev,
     case usb_redir_inval:
         WARNING("got invalid param error from usb-host?\n");
         return USB_RET_NAK;
+    case usb_redir_babble:
+        return USB_RET_BABBLE;
     case usb_redir_ioerror:
     case usb_redir_timeout:
     default:
-- 
1.7.10.4

