--- qemu-kvm-1.1.0/linux-user/signal.c	2012-06-30 03:14:15.000000000 +0100
+++ qemu-kvm-1.1.0.glibc/linux-user/signal.c	2012-07-05 11:08:36.839294266 +0100
@@ -3476,7 +3476,7 @@
 struct rt_signal_frame {
         struct siginfo *pinfo;
         void *puc;
-        struct siginfo info;
+        siginfo_t info;
         struct ucontext uc;
         uint8_t retcode[8];       /* Trampoline code. */
 };
