diff -up qemu-kvm-devel-85/qemu/Makefile.add-cache-utils-to-block-objs qemu-kvm-devel-85/qemu/Makefile
--- qemu-kvm-devel-85/qemu/Makefile.add-cache-utils-to-block-objs	2009-04-27 15:54:16.000000000 +0100
+++ qemu-kvm-devel-85/qemu/Makefile	2009-04-27 15:53:51.000000000 +0100
@@ -49,7 +49,7 @@ recurse-all: $(SUBDIR_RULES)
 #######################################################################
 # BLOCK_OBJS is code used by both qemu system emulation and qemu-img
 
-BLOCK_OBJS=cutils.o qemu-malloc.o
+BLOCK_OBJS=cutils.o cache-utils.o qemu-malloc.o
 BLOCK_OBJS+=block-cow.o block-qcow.o aes.o block-vmdk.o block-cloop.o
 BLOCK_OBJS+=block-dmg.o block-bochs.o block-vpc.o block-vvfat.o
 BLOCK_OBJS+=block-qcow2.o block-parallels.o block-nbd.o
