diff -up qemu-kvm-devel-85/qemu/configure.disable-preadv qemu-kvm-devel-85/qemu/configure
--- qemu-kvm-devel-85/qemu/configure.disable-preadv	2009-04-27 16:09:05.000000000 +0100
+++ qemu-kvm-devel-85/qemu/configure	2009-04-27 16:09:36.000000000 +0100
@@ -1189,9 +1189,9 @@ cat > $TMPC <<EOF
 int main(void) { preadv; }
 EOF
 preadv=no
-if $cc $ARCH_CFLAGS -o $TMPE $TMPC > /dev/null 2> /dev/null ; then
-  preadv=yes
-fi
+#if $cc $ARCH_CFLAGS -o $TMPE $TMPC > /dev/null 2> /dev/null ; then
+#  preadv=yes
+#fi
 
 ##########################################
 # fdt probe
