--- qemu-0.6.1/Makefile.strip	2005-02-13 14:38:23.000000000 +0000
+++ qemu-0.6.1/Makefile	2005-02-13 14:42:27.000000000 +0000
@@ -1,6 +1,6 @@
 -include config-host.mak
 
-CFLAGS=-Wall -O2 -g -fno-strict-aliasing 
+CFLAGS=$(RPM_OPT_FLAGS) -Wall -fno-strict-aliasing 
 ifdef CONFIG_DARWIN
 CFLAGS+= -mdynamic-no-pic
 endif
@@ -45,7 +45,7 @@
 install: all 
 	mkdir -p "$(bindir)"
 ifndef CONFIG_WIN32
-	install -m 755 -s $(TOOLS) "$(bindir)"
+	install -m 755 $(TOOLS) "$(bindir)"
 endif
 	mkdir -p "$(datadir)"
 	install -m 644 pc-bios/bios.bin pc-bios/vgabios.bin \
--- qemu-0.6.1/Makefile.target.strip	2005-02-13 14:38:19.000000000 +0000
+++ qemu-0.6.1/Makefile.target	2005-02-13 14:42:49.000000000 +0000
@@ -7,7 +7,7 @@
 VPATH+=:$(SRC_PATH)/linux-user
 DEFINES+=-I$(SRC_PATH)/linux-user -I$(SRC_PATH)/linux-user/$(TARGET_ARCH)
 endif
-CFLAGS=-Wall -O2 -g -fno-strict-aliasing
+CFLAGS=$(RPM_OPT_FLAGS) -Wall -fno-strict-aliasing
 LDFLAGS=-g
 LIBS=
 HELPER_CFLAGS=$(CFLAGS)
@@ -148,7 +148,7 @@
 ifeq ($(ARCH),alpha)
 # -msmall-data is not used because we want two-instruction relocations
 # for the constant constructions
-OP_CFLAGS=-Wall -O2 -g
+OP_CFLAGS=$(RPM_OPT_FLAGS) -Wall
 # Ensure there's only a single GP
 CFLAGS += -msmall-data
 LDFLAGS+=-Wl,-T,$(SRC_PATH)/alpha.ld
@@ -397,7 +397,7 @@
 
 install: all 
 ifneq ($(PROGS),)
-	install -m 755 -s $(PROGS) "$(bindir)"
+	install -m 755 $(PROGS) "$(bindir)"
 endif
 
 ifneq ($(wildcard .depend),)
