diff -urNp qemu-0.9.1.orig/Makefile.target qemu-0.9.1/Makefile.target
--- qemu-0.9.1.orig/Makefile.target	2009-01-11 23:01:20.000000000 +0530
+++ qemu-0.9.1/Makefile.target	2009-01-11 23:02:16.000000000 +0530
@@ -632,7 +632,7 @@ clean:
 
 install: all
 ifneq ($(PROGS),)
-	$(INSTALL) -m 755 -s $(PROGS) "$(DESTDIR)$(bindir)"
+	$(INSTALL) -m 755 $(PROGS) "$(DESTDIR)$(bindir)"
 endif
 
 ifneq ($(wildcard .depend),)
